### PR TITLE
nxproxy: Support communication over stdin. Allows nxcl to use standard ssh instead of nxssh.

### PR DESCRIPTION
**Pull request by sunweaver**
Patch from Gentoo. Ported 1:1.

### CODE REVIEW

**Comment by sunweaver** [comment]
@voyageur: I am about to port one of your Gentoo patch to nx-libs / nxproxy. Could you please shortly visit us here and confirm that the patch is still valid and needed? Thanks!

**Comment by voyageur** [comment]
Hmm, this Arctica project looks interesting!
This patch origin is here:
https://mail.kde.org/pipermail/freenx-knx/2008-March/006794.html
https://mail.kde.org/pipermail/freenx-knx/2008-April/006975.html

As far as I remember, only nxcl/qtnx needed it to work properly, other nx servers at the time did fine without the patch,  as still does x2go upstream I think). So it should not be useful to keep these days (I'll make a not for the next Gentoo package bump to test without it)

**Comment by sunweaver** [comment]
Hi Bernhard,

On  Mo 06 Feb 2017 22:58:00 CET, Bernard Cafarelli wrote:

> Hmm, this Arctica project looks interesting!
> This patch origin is here:
> https://mail.kde.org/pipermail/freenx-knx/2008-March/006794.html
> https://mail.kde.org/pipermail/freenx-knx/2008-April/006975.html
>
> As far as I remember, only nxcl/qtnx needed it to work properly,  
> other nx servers at the time did fine without the patch,  as still  
> does x2go upstream I think). So it should not be useful to keep  
> these days (I'll make a not for the next Gentoo package bump to test  
> without it)

Thanks for your feedback. There is another use case of NX that comes  
from devs who like stdin communications. Let's see what they think!

@vatral: @salva: any use case for nxproxy -stdin support from your  
side? Otherwise, we will drop the Gentoo patch and close this issue.

Thanks+Greets,
Mike

-- 

DAS-NETZWERKTEAM
mike gabriel, herweg 7, 24357 fleckeby
mobile: +49 (1520) 1976 148
landline: +49 (4354) 8390 139

GnuPG Fingerprint: 9BFB AEE8 6C0A A5FF BF22  0782 9AF4 6B30 2577 1B31
mail: mike.gabriel@das-netzwerkteam.de, http://das-netzwerkteam.de

**Comment by salva** [comment]
From the [QVD](https://github.com/theqvd/theqvd/) side, this is something we would really like to see incorporated into `nxproxy`, as we try to avoid opening ports on localhost as much as possible (at least when the OS allows it).

@sunweaver, feel free to assign the issue to @vatral if you thing it needs further work or just for reviewing/testing.

**Comment by sunweaver** [comment]
@vatral: Please review this PR (#317) and if possible, try to provide a test script later. Thanks.

**Comment by sunweaver** [comment]
@vatral: @salva: If you want to this PR #317 merged into nx-libs 3.6.x, I'd appreciate a code review. Ideally plus some tests. Thanks!

**Comment by sunweaver** [comment]
@vatral: @salva: ping on #317. Code review is needed if you want to see this PR merged. Thanks!

**Comment by vatral** [comment]
Hello,

At the moment I'm busy with an urgent matter, but I will take a look at that before the weekend.

File: nxproxy/Main.c [hunk 1]
```diff
@@ -49,15 +51,33 @@ int main(int argc, const char **argv)
...
   }
 
-  /*
-   * This should not return...
-   */
+  if ((nx_commfd_str = getenv("NX_COMMFD")) != NULL)
+  {
+    int nx_commfd = atoi(nx_commfd_str);
+
+    if (nx_commfd)
```
Comment (vatral) [inline nxproxy/Main.c]: STDIN is filehandle 0, and will evaluate as false in this comparison.

**Review by vatral — CHANGES_REQUESTED** [review]
The stdin issue needs fixing.

I'm looking at how to do a test script for this.

**Comment by sunweaver** [comment]
@vatral: next iteration. I actually did some testing of the patch myself now and had to rearrange some stuff in nxproxy's Main.c compared to the previous approach. Please look again. Thanks!

**Review by vatral — APPROVED** [review]

**Comment by uli42** [comment]
I am wondering how to use this? Is there any example how to establish an connection to the nxagent with this method?

**Pull request merged**